### DISCUSSION TITLE
Distributed training with train vocoder gan py error

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#365] [Archived]
<USER>>>> dyson<br/> <DATE> [March 4, 2021, 12:04pm]<br/> <br/> Hello everyone,<br/> <br/> So far, I've successfully trained a model with Tacotron 2 and the<br/> synthesized speech with Universal FullBand-MelGAN sounds okay. To<br/> further improve the quality and makes the synthesized voice sounds more<br/> similar to the original speaker, I decided to train my own vocoder using<br/> the same dataset. *slash* <br/> But when I use the following command: *slash* <br/> `CUDA_VISIBLE_DEVICES='0,1,2' OMP_NUM_THREADS=1 python TTS/bin/distribute.py --script train_vocoder_gan.py --config_path config_vocoder_PWgan.json`<br/> <br/> I got the following output:<br/> <br/> Traceback (most recent call last):<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 654, in <module><br/> main(args)<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 559, in main<br/> epoch)<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 114, in train<br/> y_hat = model_G(c_G)<br/> File '/home/ldai/anaconda3/envs/mozillatts/lib/python3.6/site-packages/torch/nn/modules/module.py', line 727, in _call_impl<br/> result = self.forward(*input, **kwargs)<br/> File '/home/ldai/anaconda3/envs/mozillatts/lib/python3.6/site-packages/torch/nn/parallel/distributed.py', line 606, in forward<br/> if self.reducer._rebuild_buckets():<br/> RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).<br/> <br/> I didn't get such an error when using train_tacotron.py with the same<br/> command. *slash* <br/> Any suggestions?<br/> <br/> Libraries version: *slash* <br/> python=3.6.12 *slash* <br/> torch=1.7.1<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/distributed-training-with-train-vocoder-gan-py-error]<br/>

**Comment by JRMeyer** [comment]
## [Archived] Distributed training with train vocoder gan py error<br/> <br/> <USER>>>> dyson<br/> <DATE> [March 4, 2021, 12:04pm]<br/> <br/> Hello everyone,<br/> <br/> So far, I've successfully trained a model with Tacotron 2 and the<br/> synthesized speech with Universal FullBand-MelGAN sounds okay. To<br/> further improve the quality and makes the synthesized voice sounds more<br/> similar to the original speaker, I decided to train my own vocoder using<br/> the same dataset. *slash* <br/> But when I use the following command: *slash* <br/> `CUDA_VISIBLE_DEVICES='0,1,2' OMP_NUM_THREADS=1 python TTS/bin/distribute.py --script train_vocoder_gan.py --config_path config_vocoder_PWgan.json`<br/> <br/> I got the following output:<br/> <br/> Traceback (most recent call last):<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 654, in <module><br/> main(args)<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 559, in main<br/> epoch)<br/> File '/home/ldai/projects/TTS/TTS/bin/train_vocoder_gan.py', line 114, in train<br/> y_hat = model_G(c_G)<br/> File '/home/ldai/anaconda3/envs/mozillatts/lib/python3.6/site-packages/torch/nn/modules/module.py', line 727, in _call_impl<br/> result = self.forward(*input, **kwargs)<br/> File '/home/ldai/anaconda3/envs/mozillatts/lib/python3.6/site-packages/torch/nn/parallel/distributed.py', line 606, in forward<br/> if self.reducer._rebuild_buckets():<br/> RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).<br/> <br/> I didn't get such an error when using train_tacotron.py with the same<br/> command. *slash* <br/> Any suggestions?<br/> <br/> Libraries version: *slash* <br/> python=3.6.12 *slash* <br/> torch=1.7.1<br/> <br/> ### This is an archived TTS discussion thread from discourse.mozilla.org/t/distributed-training-with-train-vocoder-gan-py-error<br/>

**Reply by OmiWakode** [reply]
I also got a similar error while multi gpu training of the custom vits model. Did you solve this error?
```
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/train_tts_env/lib/python3.8/site-packages/trainer/trainer.py", line 1461, in fit
    self._fit()
  File "/home/ubuntu/anaconda3/envs/train_tts_env/lib/python3.8/site-packages/trainer/trainer.py", line 1449, in _fit
    self.test_run()
  File "/home/ubuntu/anaconda3/envs/train_tts_env/lib/python3.8/site-packages/trainer/trainer.py", line 1385, in test_run
    self.model.test_log(test_outputs, self.dashboard_logger, self.training_assets, self.total_steps_done)
  File "/home/ubuntu/anaconda3/envs/train_tts_env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'DistributedDataParallel' object has no attribute 'test_log'
```
